File: 프로그래머스/2/250136. ［PCCP 기출문제］ 2번 ／ 석유 시추/［PCCP 기출문제］ 2번 ／ 석유 시추.cpp
#include <string>
#include <vector>
#include<iostream>
#include<set>

using namespace std;
int land_copy[501][501],area[501][501], number,cnt;
int dx[4]={1,-1,0,0};
int dy[4]={0,0,1,-1};
vector<int>areaSize;

void go(int x, int y){
    area[x][y]=number;
    for(int i=0;i<4;i++){
        int nx = x + dx[i];
        int ny = y + dy[i];
        if(land_copy[nx][ny]&&area[nx][ny]==-1){
            cnt++;
            go(nx, ny);
        }
    }
}

int solution(vector<vector<int>> land) {
    int answer = 0;
    int N = land.size(),M = land[0].size();
    for(int i=0;i<N;i++)
        for(int j=0;j<M;j++){
            area[i][j]=-1;
            land_copy[i][j]=land[i][j];
        }
    for(int i=0;i<N;i++){
        for(int j=0;j<M;j++){
            if(land[i][j]&&area[i][j]==-1){
                cnt = 1;
                go(i,j);
                number++;
                areaSize.push_back(cnt);
            }
        }
    }
    for(int j=0;j<M;j++){
        int sum = 0;
        set<int>visited;
        for(int i=0;i<N;i++){
            if(area[i][j]!=-1){
                if(visited.find(area[i][j])==visited.end()){
                    sum+=areaSize[area[i][j]];
                    visited.insert(area[i][j]);
                }
            }
        }
        answer = max(answer,sum);
    }
    return answer;
}

File: 프로그래머스/2/250136. ［PCCP 기출문제］ 2번 ／ 석유 시추/［PCCP 기출문제］ 2번 ／ 석유 시추.java
import java.util.*;
class Solution {
    
    public int cnt = 0; 
    public int number = 0;
    public int[] dx = {1,-1,0,0};
    public int[] dy = {0,0,-1,1};
    public int[][]area;
    public int[][] landCopy;
    public List<Integer> areaSize;
    
    public int solution(int[][] land) {
        int answer = 0;
        area = new int[land.length][land[0].length];
        landCopy = new int[land.length][land[0].length];
        areaSize = new ArrayList<>();
        for(int i=0;i<land.length;i++){
            for(int j=0;j<land[i].length;j++){
                area[i][j] = -1;
                landCopy[i][j] = land[i][j];
            }
        }
        for(int i=0;i<land.length;i++){
            for(int j=0;j<land[i].length;j++){
                if(land[i][j]==1 && area[i][j]==-1){
                    cnt = 1;
                    go(i,j);
                    number+=1;
                    areaSize.add(cnt);
                }
            }
        }
        for(int j=0;j<land[0].length;j++){
            Set<Integer> visited = new HashSet<>();
            int sum = 0;
            for(int i=0;i<land.length;i++){
                if(area[i][j]!=-1){
                    if(!visited.contains(area[i][j])){
                        sum +=areaSize.get(area[i][j]);
                        visited.add(area[i][j]);
                    }
                }
            }
            answer = Math.max(answer,sum);
        }
        return answer;
    }
    
    public void go(int x, int y){
        area[x][y]=number;
        for(int i=0;i<4;i++){
            int nx = x + dx[i];
            int ny = y + dy[i];
            if(nx>=0&&nx<landCopy.length&&ny>=0&&ny<landCopy[0].length){
                if(landCopy[nx][ny]==1 && area[nx][ny]==-1){
                    cnt+=1;
                    go(nx, ny);
                }
            }
        }
    }   
}